accepted/0000-handling-peer-conflicts.md

```diff
@@ -0,0 +1,337 @@
+  the conflicted version.
+- **Non-strict mode**: Raise an error and halt the ideal tree building
+  process, as no adequate resolution can be inferred.
+
+#### Peer Set Conflicts with Non-Peer Dependency in Tree
+
+It is possible for a peer set to conflict with an existing non-peer
+dependency in the tree in the only place where it can be placed.  In these
+cases, we prioritize the non-peer dependency over the peers in the peer
+set.
+
+Example 1:
+
+```
+root -> (x@1, y@2)
+x@1 -> PEER(y@1)
+```
+
+In this case, the root project depends on `x`, which has a peer dependency
+on `y@1`, which conflicts with the root project's non-peer dependency on
+`y@2`.
+
+Example 2:
+
+```
+root -> (x@1, z@2)
+x@1 -> (y@1)
+y@1 -> PEER(z@1)
+```
+
+Note that this example does _not_ cause a conflict, because the following
+arrangement correctly meets the dependency contracts:
+
+```
+root
++-- x@1
+|   +-- y@1
+|   +-- z@1
++-- z@2
+```
+
+However, it is important to note the conflict with the peer set when
+attempting to place the `(y@1, z@1)` peer set, so that it can be
+placed deeper in the tree, underneath `x@1`.
+
+When a conflict cannot be resolved correctly, as in Example 1 above, do the
+following:
+
+- **Strict mode**: Raise an error and halt the ideal tree building process.
+- **Force mode**: Warn about the conflict.  Skip the conflicted peer
+  dependency, and keep the non-peer dependency.
+- **Non-strict mode**: If the source of the conflicted non-peer dependency
+  is the root project, then raise an error and halt the process.  If it is
+  not, then raise a warning about the conflict and proceed as with force
+  mode.
+
+#### Peer Set Conflicts with Peer Dependency in Tree
+
+When we do not have a non-peer dependency involved in the peer conflict,
+the situation is impossible to resolve appropriately.  (This is similar to
+the **Conflict Internal to the Peer Set** case above.)
+
+Example 1:
+
+```
+project -> (x@1, y@2)
+x@1 -> PEER(z@1)
+y@2 -> PEER(z@2)
+```
+
+In this case, the only place we can put the dependencies is a location
+where the peer dependencies will conflict with one another.  If `project`
+is not the root project, but instead a dependency, it is _possible_ that it
+might be resolvable via duplication.  For example:
+
+```
+root
++-- project
+|   +-- x@1
+|   +-- z@1
++-- y@2
++-- z@2
+```
+
+However, there are situations where this is not the case.
+
+Example 2:
+
+```
+root -> (project, y@3, x@3)
+project -> (x@1, y@2)
+x@1 -> PEER(z@1)
+x@3 -> PEER(z@3)
+y@2 -> PEER(z@2)
+y@3 -> PEER(z@3)
+```
+
+In this case, the root project will require that `x`, `y`, and `z` all are
+in the first `node_modules` folder at version 3.  This leaves `project` as
+the only place where its dependencies can be placed, causing the conflict.
+
+- **Strict mode**: Raise an error and halt the ideal tree building process.
+- **Force mode**: Warn about the conflict.  Arbitrarily keep the first peer
```

Comment:
would this "first peer dependency" happen to be the first one declared in the `package.json` file? Otherwise if there's no way at all to define it, it seem to me it might be better to just raise an error here too.

Reply:
It would be the first one encountered in a breadth-first traversal through the dependency graph, ordered alphabetically.